sstable/src/block_reader.rs

```diff
@@ -36,19 +34,42 @@ impl<'a> BlockReader<'a> {
 
     pub fn read_block(&mut self) -> io::Result<bool> {
         self.offset = 0;
-        let block_len_res = read_u32(self.reader.as_mut());
-        if let Err(err) = &block_len_res {
-            if err.kind() == io::ErrorKind::UnexpectedEof {
-                return Ok(false);
+        self.buffer.clear();
+
+        let block_len = match self.reader.len() {
+            0 => return Ok(false),
+            1..=3 => {
+                return Err(io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "failed to read block_len",
+                ))
             }
-        }
-        let block_len = block_len_res?;
-        if block_len == 0u32 {
-            self.buffer.clear();
+            _ => self.reader.read_u32(),
+        };
+        if block_len == 0 {
             return Ok(false);
         }
-        self.buffer.resize(block_len as usize, 0u8);
-        self.reader.read_exact(&mut self.buffer[..])?;
+
+        let real_block_len = (block_len & !COMPRESSION_FLAG) as usize;
+        if self.reader.len() < real_block_len {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "failed to read block content",
+            ));
+        }
+        if real_block_len as u32 != block_len {
+            let required_capacity =
+                Decompressor::upper_bound(&self.reader[..real_block_len]).unwrap_or(1024 * 1024);
+            self.buffer.reserve(required_capacity);
+            Decompressor::new()?
```

Comment:
Not sure whether the difference is measurable, but many you can save some initialization overhead by moving the `Decompressor` into the `BlockReader` and re-using it for each call to `read_block`? (And similarly for the `Compressor` but I guess this might even be less significant in the compression path.)

Reply:
I already tried it and it made no difference whatsoever. I think it as more of an impact when using a dictionary